### DISCUSSION TITLE
New datasource: AWS Systems Manager Parameter Store

### DISCUSSION THREAD

**Discussion by izeau** [renovatebot/renovate#25852] [Suggest an Idea]
## Use cases

This feature request fills a niche for at least a couple use cases: the retrieval of the latest versions of Amazon Linux AMI and AWS for Fluent Bit images. Last but not least, it may be used in a gitops workflow, where developers publish parameters when they update an application / library or publish a new OCI image, and renovate automatically picks them up to update the infrastructure.

**Amazon Linux AMI**

AWS regularly publishes updates to their Amazon Linux images. Such images can be discovered using the `aws-machine-image` experimental datasource, however the `DescribeImages` API filtering is rather basic and only kind of works if the publisher uses an easy-to-match and non-ambiguous format. This can be an issue, for instance the `al2023-ami-*-x64_86` name filter returns several variants of the image, the most recent of which are:
- al2023-ami-2023.2.20231113.0-kernel-6.1-x86_64
- al2023-ami-ecs-hvm-2023.0.20231103-kernel-6.1-x86_64
- al2023-ami-ecs-neuron-hvm-2023.0.20231103-kernel-6.1-x86_64
- al2023-ami-minimal-2023.2.20231113.0-kernel-6.1-x86_64

However, AWS also publishes parameters to retrieve the AMI in the AWS SSM Parameter Store, for instance:
- /aws/service/ami-amazon-linux-latest/al2023-ami-kernel-default-x86_64
- /aws/service/ami-amazon-linux-latest/al2023-ami-minimal-kernel-default-x86_64
- /aws/service/ecs/optimized-ami/amazon-linux-2023/neuron/recommended
- /aws/service/ecs/optimized-ami/amazon-linux-2023/recommended

Using these parameters provides straightforward retrieval of the latest AMI for a given variant:

```shell
$ aws ssm get-parameter --name /aws/service/ami-amazon-linux-latest/al2023-ami-kernel-default-x86_64
{
    "Parameter": {
        "Name": "/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-default-x86_64",
        "Type": "String",
        "Value": "ami-07355fe79b493752d",
        "Version": 58,
        "LastModifiedDate": "2023-11-14T21:17:19.573000+01:00",
        "ARN": "arn:aws:ssm:eu-west-1::parameter/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-default-x86_64",
        "DataType": "text"
    }
}
```

**AWS for Fluent Bit images**

AWS [recommends using SSM](https://github.com/aws/aws-for-fluent-bit#using-ssm-to-find-available-versions-and-aws-regions) to find the recommended stable `aws-for-fluent-bit` image version:

```shell
$ aws ssm get-parameter --name /aws/service/aws-for-fluent-bit/stable
{
    "Parameter": {
        "Name": "/aws/service/aws-for-fluent-bit/stable",
        "Type": "String",
        "Value": "906394416424.dkr.ecr.eu-west-1.amazonaws.com/aws-for-fluent-bit:2.31.12.20231011",
        "Version": 16,
        "LastModifiedDate": "2023-11-14T00:46:38.482000+01:00",
        "ARN": "arn:aws:ssm:eu-west-1::parameter/aws/service/aws-for-fluent-bit/stable",
        "DataType": "text"
    }
}
```

## Implementation idea

I would love to submit a PR for an `aws-parameter-store` datasource that uses the AWS SDK to fetch the value of a given parameter, either using its name or its ARN:

```hcl
resource "aws_launch_template" "web" {
  # parameter=/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-default-x86_64 depName=amazon-linux-2023 versioning=aws-machine-image
  ami = "ami-07355fe79b493752d"
}
```

Using the parameter ARN allows the user to specify a different region, which can be useful for cross-region infrastructure:

```hcl
resource "aws_launch_template" "web" {
  # parameter=arn:aws:ssm:eu-west-1::parameter/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-default-x86_64 depName=amazon-linux-2023 versioning=aws-machine-image
  ami = "ami-07355fe79b493752d"
}
```

I have a working prototype which I can submit as a PR if this is a wanted feature.

Cheers!

**Comment by rarkins** [comment]
@izeau sorry for the delay in responding here

Using https://github.com/renovatebot/renovate/tree/main/lib/modules/datasource/aws-machine-image as inspiration, could you try to map these concepts to the existing "datasource" semantics/API which Renovate uses? In particular, we use `registryUrl` and `packageName`. And if it can't be mapped, which would mean we need to change this interface, can you suggest the best approaches?
